docs/proposals/23-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,347 @@
+      - ['MinChange' stabilisation prevents response to node CPU exhaustion](#minchange-stabilisation-prevents-response-to-node-cpu-exhaustion)
+      - [HPA-VPA policy fusion model does not account for multiplicative interaction](#hpa-vpa-policy-fusion-model-does-not-account-for-multiplicative-interaction)
+    - [Other Benefits to Replacing the Existing Solution](#other-benefits-to-replacing-the-existing-solution)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Rationale](#rationale)
+    - [Design Outline](#design-outline)
+    - [Element: Gardener Custom Metrics Provider Component](#element-gardener-custom-metrics-provider-component)
+    - [Element: New Custom Pod Metric for ShootKapis](#element-new-custom-pod-metric-for-shootkapis)
+    - [Element: HPA](#element-hpa)
+    - [Element: VPA](#element-vpa)
+    - [Transition Strategy](#transition-strategy)
+  - [Discussion and Limitations](#discussion-and-limitations)
+      - [Gardener-custom metrics precludes the use of other sources of custom metrics](#gardener-custom-metrics-precludes-the-use-of-other-sources-of-custom-metrics)
+      - [Observed fluctuations in actual compute efficiency affecting ShootKapi workloads](#observed-fluctuations-in-actual-compute-efficiency-affecting-shootkapi-workloads)
+  - [Alternatives](#alternatives)
+  - [References](#references)
+
+## Summary
+When it comes to autoscaling shoot control plane `kube-apiserver` instances (ShootKapi hereafter), Gardener needs
+both stability and efficiency (accurate scaling). The existing approach of fusing HPA and VPA into the 2-dimensional
+autoscaler which is HVPA, poses severe algorithmic limitations which manifest as stability and efficiency issues in the
+field. 
+
+This document outlines the main deficiencies in the existing solution and proposes a simple design which overcomes them,
+reduces maintenance effort by reducing Gardener's custom code base, and enables subsequent reduction in ShootKapi
+compute cost.
+
+## Motivation
+Gardener uses a relatively complex approach for scaling shoot control plane `kube-apiserver` instances (ShootKapi).
+This need is a consequence of the following concerns:
+1. Gardener supplies K8s clusters of highly diverse compute scale and utilization - ranging from a couple of small
+   practically idle nodes, to hundreds of heavily utilized giants. Furthermore, the ratio of 'cluster compute capacity'
+   to 'ShootKapi load' substantially varies between different applications utilizing Gardener shoots. As a result,
+   compute resources required by ShootKapis of different shoots vary by two orders of magnitude.
+2. The compute cost of ShootKapi workloads constitutes a major part of Gardener's overall compute cost. Small relative
+   waste in sizing ShootKapis results in considerable absolute cost.
+3. Kube-apiserver does not scale well horizontally, in a sense that extra replicas come with substantial compute overhead.   
+
```

Comment:
The main disadvantages I am aware of are:
- HVPA scales first horizontally (up to a magic number of 4) before it scales vertically, which adds more wasteful replicas than actually needed (every replica needs to have its caches, etc.)
- The number 4 is somewhat random (results from 1 performance test (!); assumptions that I have strong doubts about; disproven by some Gardener hosters that run seeds with 6 KAPIs and possibly even more). In addition, it fits poorly to HA (one zone running twice as many as the other zones (with 3 zones)).
- Whenever HVPA makes a decision, the entire deployment is rolled, resulting in constant waste. While @voelzmo notes that VPA will "roll a pod" at the latest after 12h, the current HVPA implementation is even more wasteful as "all pods are rolled" and all need to start from scratch to fill their caches, etc.